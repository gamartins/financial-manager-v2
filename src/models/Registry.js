export default class Registry {
  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  static STATUS = {
    NOT_PAID: 0,
    PAID: 1,
    PAID_FINANCIAL_CASH: 2,
    RESERVERD_SAVINGS: 3,
  }

  static getStatusString(status) {
    switch (status) {
    case Registry.STATUS.NOT_PAID:
      return 'NÃ£o pago'
    case Registry.STATUS.PAID:
      return 'Pago'
    case Registry.STATUS.PAID_FINANCIAL_CASH:
      return 'Depositar em poup'
    case Registry.STATUS.RESERVERD_SAVINGS:
      return 'Depositado em poup'
    default:
      return ''
    }
  }
}
