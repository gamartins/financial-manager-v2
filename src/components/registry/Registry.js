import React, { Component } from 'react'

import addMonths from 'date-fns/addMonths'

import RegistryForm from './RegistryForm'
import RegistryList from './RegistryList'
import DatePagination from '../date-pagination/DatePagination'
import RegistryService from '../../services/RegistryService'

export default class Registry extends Component {

  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  constructor(props) {
    super(props)

    this.state = {
      date: new Date(),
      dataList: [],
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: ''
      }
    }
  }

  componentDidMount() {
    this.getAll()
  }

  getAll() {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`

    const dataList = RegistryService.get(dateQuery)
    
    this.setState({ dataList })
  }

  getNext() {
    this.setState(
      (state) => ({ date: addMonths(state.date, 1) }),
      () => this.getAll())
  }

  getPrevious() {
    this.setState(
      (state) => ({ date: addMonths(state.date, -1) }),
      () => this.getAll())
  }

  setFormValue(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index === -1) {
      return
    }

    const newFormValues = { ...this.state.dataList[index] }
    
    this.setState({ formValues: newFormValues })
  }

  resetFormValue() {
    this.setState({
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: '',
      }
    })
  }

  save(values) {
    return values.id ? this.edit(values) : this.add(values)
  }

  add(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const dataList = [ ...this.state.dataList ]
    dataList.push({
      id: Math.ceil((Math.random() * 100000)),
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    })

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  edit(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const index = this.state.dataList.findIndex(item => item.id === values.id)
    const dataList = [ ...this.state.dataList ]
    dataList[index] = {
      id: values.id,
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    }

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  remove(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index !== -1) {
      this.state.dataList.splice(index, 1)
      this.setState({ dataList: this.state.dataList })
      this.resetFormValue()
    }    
  }

  render() {
    return (
      <div>
        <RegistryForm
          initialValues={this.state.formValues}
          saveItem={values => this.save(values)} />
        <DatePagination
          date={this.state.date}
          format={'MMM'}
          previous={() => this.getPrevious()}
          next={() => this.getNext()}/>
        <RegistryList
          dataList={this.state.dataList}
          removeItem={id => this.remove(id)}
          editItem={item => this.setFormValue(item)} />
      </div>
    )
  }
}
