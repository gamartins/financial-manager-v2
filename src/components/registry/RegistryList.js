import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Table from 'react-bootstrap/Table'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons'

import Registry from './Registry'
import CategoryService from '../../services/CategoryService'

export default class RegistryList extends Component {
  categoryList = []

  componentDidMount() {
    this.categoryList = CategoryService.get()
  }

  removeItem(item) {
    this.props.removeItem(item)
  }

  editItem(item) {
    this.props.editItem(item)
  }

  calcTotal() {
    const proventos = this.props.dataList
      .filter(item => item.type === Registry.TYPE.PROVENTO)
      .map(item => item.price)
      .reduce((acc, current) => acc + current, 0)

    const descontos = this.props.dataList
      .filter(item => item.type === Registry.TYPE.DESCONTO)
      .map(item => item.price)
      .reduce((acc, current) => acc + current, 0)

    return proventos - descontos
  }

  findCategory(categoryId) {
    const category = this.categoryList.find(category => category.id === categoryId)

    return category ? category.name : 'Sem categoria'
  }

  findStatus(status) {
    switch (status) {
      case Registry.STATUS.NOT_PAID:
        return 'Não pago'
      case Registry.STATUS.PAID:
        return 'Pago'
      case Registry.STATUS.PAID_FINANCIAL_CASH:
        return 'Depositar em poup'
      case Registry.STATUS.RESERVERD_SAVINGS:
        return 'Depositado em poup'
      default:
        return ''
    }
  }

  render() {
    return (
      <div>
        <Table hover>

          <thead>
            <tr>
              <th>Nome</th>
              <th>Tipo</th>
              <th>Categoria</th>
              <th>Status</th>
              <th>Valor</th>
              <th>Ações</th>
            </tr>
          </thead>

          <tbody>
            {this.props.dataList.map(item => (
              <tr key={item.id}>
                <td>{item.name}</td>
                <td>{item.type === Registry.TYPE.PROVENTO ? 'Proveto' : 'Desconto' }</td>
                <td>{this.findCategory(item.category)}</td>
                <td>{this.findStatus(item.status)}</td>
                <td>R$ {item.price}</td>
                <td>
                  <FontAwesomeIcon
                    icon={faEdit}
                    id="edit-item-icon"
                    onClick={() => this.editItem(item.id)}/>
                  <FontAwesomeIcon
                    icon={faTrash}
                    id="remove-item-icon"
                    onClick={() => this.removeItem(item.id)} />
                </td>
              </tr>
            ))}
          </tbody>

          <tfoot>
            <tr>
              <td colSpan="4">Total</td>
              <td id='cell-total' colSpan="2">R$ {this.calcTotal()}</td>
            </tr>
          </tfoot>

        </Table>
      </div>
    )
  }
}

RegistryList.propTypes = {
  editItem: PropTypes.func,
  removeItem: PropTypes.func,
  dataList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      type: PropTypes.number,
      category: PropTypes.number,
      status: PropTypes.number,
      price: PropTypes.number,
    })
  ).isRequired
}
