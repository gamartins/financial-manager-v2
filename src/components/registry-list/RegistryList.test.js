import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import RegistryList from './RegistryList'
import Registry from '../../models/Registry'
import registerIcons from '../../services/FontAwesome'

Enzyme.configure({ adapter: new Adapter() })

describe('<RegistryList/>', () => {
  beforeEach(() => {
    registerIcons()
  })

  test('should be created', () => {
    const dataList = []
    const wrapper = mount(<RegistryList dataList={dataList} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should receive a registry list as props', () => {
    const dataList = []
    const wrapper = shallow(<RegistryList dataList={dataList} />)
    expect(wrapper.instance().props.dataList).toHaveLength(0)

    dataList.push({ id: 1 })
    expect(wrapper.instance().props.dataList).toHaveLength(1)
  })

  test('should show a table with a registry list', () => {
    const dataList = [
      { id: 1, name: 'Item 01', price: 50, type: Registry.TYPE.DESCONTO, category: 1 },
      { id: 1, name: 'Item 01', price: 50, type: Registry.TYPE.DESCONTO, category: 1 },
      { id: 1, name: 'Item 01', price: 50, type: Registry.TYPE.DESCONTO, category: 1 },
    ]

    const wrapper = shallow(<RegistryList dataList={dataList} />)
    expect(wrapper.find('tbody tr')).toHaveLength(3)
  })

  test('should call method removeItem on delete icon click', () => {
    const dataList = [{ id: 1 }]
    const onDeleteMock = jest.fn(() => {})

    const wrapper = shallow(<RegistryList dataList={dataList} removeItem={onDeleteMock} />)
    wrapper.find('#remove-item-icon').simulate('click')

    expect(onDeleteMock).toBeCalled()
    expect(onDeleteMock).toBeCalledWith(1)
  })

  test('should call method editItem on edit icon click', () => {
    const dataList = [{ id: 1 }]
    const onEditFunction = jest.fn(() => {})

    const wrapper = shallow(<RegistryList dataList={dataList} editItem={onEditFunction} />)
    wrapper.find('#edit-item-icon').simulate('click')

    expect(onEditFunction).toBeCalled()
    expect(onEditFunction).toBeCalledWith(1)
  })

  test('should show a balance based on item list', () => {
    const dataList = [
      { id: 1, name: 'Item 01', price: 1000, type: Registry.TYPE.PROVENTO, category: 1 },
      { id: 2, name: 'Item 02', price: 50, type: Registry.TYPE.DESCONTO, category: 1 },
      { id: 3, name: 'Item 03', price: 50, type: Registry.TYPE.DESCONTO, category: 1 },
    ]

    const wrapper = shallow(<RegistryList dataList={dataList} />)
    expect(wrapper.find('#cell-total').text()).toEqual('R$ 900')
  })

  test('should subtract values of Registry.TYPE.DESCONTO', () => {
    const dataList = [{ id: 1, name: 'Item 01', price: 1000, type: Registry.TYPE.PROVENTO, category: 1 }]
    const wrapper = shallow(<RegistryList dataList={dataList} />)

    expect(wrapper.find('#cell-total').text()).toEqual('R$ 1000')

    dataList.push({ id: 2, name: 'Item 02', price: 50, type: Registry.TYPE.DESCONTO, category: 1 })
    wrapper.setState({ dataList })

    expect(wrapper.find('#cell-total').text()).toEqual('R$ 950')
  })

  test('should add values of Registry.TYPE.PROVENTO', () => {
    const dataList = [{ id: 1, name: 'Item 01', price: 1000, type: Registry.TYPE.PROVENTO, category: 1 }]
    const wrapper = shallow(<RegistryList dataList={dataList} />)

    expect(wrapper.find('#cell-total').text()).toEqual('R$ 1000')

    dataList.push({ id: 2, name: 'Item 02', price: 50, type: Registry.TYPE.PROVENTO, category: 1 })
    wrapper.setState({ dataList })

    expect(wrapper.find('#cell-total').text()).toEqual('R$ 1050')
  })
})
