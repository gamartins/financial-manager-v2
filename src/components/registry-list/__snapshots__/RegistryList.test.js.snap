// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegistryList/> should be created 1`] = `ReactWrapper {}`;
